### PR TITLE
[Common] Support implementing PRTv3 Protocol 

### PR DESCRIPTION
**Pull request by mohitc1**
Changes to support implement PRTv3 Protocol in broker

Changes here include:
1. Added JWTClaims to MicrosoftStsTokenRequest and supporting changes in AbstractJwtRequest.
2. Added IAcquireMicrosoftStsTokenStrategy, implementations should use MicrosoftStsTokenRequest, MicrosoftStsTokenResponse.
Refer https://github.com/AzureAD/ad-accounts-for-android/pull/2167

### CODE REVIEW

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/MicrosoftTokenResponse.java [hunk 1]
```diff
@@ -91,8 +91,8 @@ public String getSessionKeyJwe() {
      * @param sesionKey the session key JWE associated with this result, or null if none.
      * @return
      */
-    public String setSessionKeyJwe(String sesionKey) {
-        return mSessionKeyJwe;
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/MicrosoftTokenResponse.java:95]: The old code here looks very strange 😁

File: common4j/src/test/com/microsoft/identity/common/java/jwt/JwtRequestTests.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package com.microsoft.identity.common.java.jwt;
```
Comment (shahzaibj) [inline common4j/src/test/com/microsoft/identity/common/java/jwt/JwtRequestTests.java:23]: LICENSE

File: common4j/src/test/com/microsoft/identity/common/java/jwt/JwtRequestTests.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.util.Map;
+
+/**
+ * Tests for {@link JwtRequestHeader}, {@link JwtRequestBody}, {@link AbstractJwtRequest}
```
Comment (shahzaibj) [inline common4j/src/test/com/microsoft/identity/common/java/jwt/JwtRequestTests.java]:
```suggestion
 * Tests for {@link JwtRequestHeader}, {@link JwtRequestBody}, {@link AbstractJwtRequest}.
```

**Review by shahzaibj — COMMENTED** [review]
Ultimately, PRTv3 is supposed to be specific to Broker only or would we also use it in MSAL as well?

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/MicrosoftTokenResponse.java (hunk 1 above)
Comment (mohitc1) [reply common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/MicrosoftTokenResponse.java:95]: yes, seems was written this way by mistake. It was never used before so remained uncaught.

**Comment by mohitc1** [comment]
> Ultimately, PRTv3 is supposed to be specific to Broker only or would we also use it in MSAL as well?

It is specific to broker only. Some of the stuff which helps PRT implementation but not specific to PRT is placed in common.

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java [hunk 4]
```diff
@@ -119,6 +83,52 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
...
+                );
+
+            case USER_CANCEL:
+                return new UserCancelException();
+            default:
+                Logger.warn(
+                        methodTag,
+                        "No AuthorizationResult status set"
+                );
+                return new ClientException(ClientException.UNKNOWN_ERROR, "Authorization failed with unknown error");
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java]: We should include the error & errorDescription information from the response in the exception here, or is that passed in some other form to the caller?

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java [hunk 5]
```diff
@@ -299,6 +309,18 @@ public static BaseException baseExceptionFromException(final Throwable exception
...
             e = exception.getCause();
         }
+        if (e instanceof BaseException) {
+            return (BaseException) e;
+        }
+
+        return clientExceptionFromException(e);
+    }
+
+    public static ClientException clientExceptionFromException(final Throwable exception) {
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java]: NonNull?

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/IAcquireMicrosoftStsTokenStrategy.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import com.microsoft.identity.common.java.commands.parameters.SilentTokenCommandParameters;
+import com.microsoft.identity.common.java.commands.parameters.TokenCommandParameters;
+import com.microsoft.identity.common.java.exception.BaseException;
+import com.microsoft.identity.common.java.exception.ClientException;
+import com.microsoft.identity.common.java.providers.oauth2.TokenResult;
+
+import lombok.NonNull;
+
+/**
+ * Acquire token for strategy for Microsoft STS
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/IAcquireMicrosoftStsTokenStrategy.java]: nit: Acquire Token Strategy

File: common4j/src/main/com/microsoft/identity/common/java/providers/oauth2/TokenRequest.java [hunk 7]
```diff
@@ -256,6 +256,7 @@ public static class GrantTypes {
         public static final String PASSWORD = "password";
         public static final String DEVICE_CODE = "urn:ietf:params:oauth:grant-type:device_code";
         public final static String CLIENT_CREDENTIALS = "client_credentials";
+        public static final String JWT_BEARER = "urn:ietf:params:oauth:grant-type:jwt-bearer";
```
Comment (iamgusain) [inline common4j/src/main/com/microsoft/identity/common/java/providers/oauth2/TokenRequest.java:261]: Can you add a comment here to describe how/when this value is used in the protocol

**Review by iamgusain — APPROVED** [review]
Can you update title/description of the PR, describing what is contained in this change.
It makes it really useful in the future to look back at the changes.

**Review by rpdome — APPROVED** [review]

**Pull request merged**